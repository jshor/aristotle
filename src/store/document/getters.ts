import { DocumentState } from "./state"

/**
 * Document zoom level.
 */
export function zoom (state: DocumentState) {
  return state.zoomLevel
}

export function totalSelectionCount (state: DocumentState) {
  return state.selectedItemIds.size +
    state.selectedConnectionIds.size +
    Object.keys(state.selectedControlPoints).length
}

export function canDelete (state: DocumentState) {
  return state.selectedItemIds.size > 0 ||
    state.selectedConnectionIds.size > 0 ||
    Object.keys(state.selectedControlPoints).length > 0
}
