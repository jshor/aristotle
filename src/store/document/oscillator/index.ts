import Pulse from '@/types/interfaces/Pulse'
import BinaryWavePulse from './BinaryWavePulse'
import Oscillogram from '@/types/types/Oscillogram'
import Point from '@/types/interfaces/Point'

/**
 * @class Oscillator
 * @description Manages all oscillating services, including binary waves and clock pulses.
 * This service uses system time to determine proper sequencing and updating.
 */
export default class Oscillator {
  /** Whether or not the oscillator is paused. */
  public isPaused: boolean = true

  /** Key-value pair of all waves. */
  public waves: Record<string, Pulse> = {}

  /** The refresh rate for the oscilloscope. */
  public refreshRate: number = 20 // TODO: this can be configured to make the waves wider/oscilloscope run faster

  /** The last time the oscillator was updated, in milliseconds. */
  public lastUpdateTime: number = Date.now()

  /** Elapsed total time, in milliseconds. */
  public timeMsElapsed: number = 0

  /** Offset wave time, in milliseconds. */
  public timeMsOffset: number = 0

  /** Change event on oscillogram generation. */
  public onTick?: (oscillogram: Oscillogram) => void

  /**
   * Starts the oscillator.
   */
  start = () => {
    if (!this.isPaused) return

    // this is the time that has elapsed prior to the completion of a full period
    // when the oscillator is stopped, we want to maintain the time progressed before the next period
    const remainingPeriodTime = this.lastUpdateTime % this.refreshRate

    this.lastUpdateTime = Date.now() + remainingPeriodTime
    this.isPaused = false
    this.tick()
  }

  /**
   * Stops the oscillator.
   */
  stop = () => {
    this.isPaused = true
  }

  /**
   * Clears all current binary wave geometries.
   */
  clear = () => {
    Object
      .values(this.waves)
      .forEach(wave => {
        if (wave instanceof BinaryWavePulse) {
          (wave as BinaryWavePulse).reset()
        }
      })
  }

  /**
   * Computes the SVG points for each given binary wave.
   *
   * @param waves
   * @returns oscillogram result
   */
  computeWaveGeometry = (waves: Record<string, Pulse>) => {
    const displays: Oscillogram = {}
    const getPoints = (segments: Point[]) => segments.map(({ x, y }) => `${x},${y}`)

    for (const id in waves) {
      if (waves[id] instanceof BinaryWavePulse) {
        const wave = waves[id] as BinaryWavePulse
        const points = getPoints(wave.segments)

        displays[id] = {
          points: points.join(' '),
          width: wave.width,
          hue: wave.hue
        }
      }
    }

    return displays
  }

  /**
   * Broadcasts the oscillogram.
   *
   * @emits change containing the oscillogram data
   */
  broadcast = () => {
    // set the max history to be twice the size of the user's screen, then cut half of it when it is exceeded
    const screenWidth = window.screen.width
    const widthOfOneSecond = 1000 / this.refreshRate
    const maxHistorySeconds = screenWidth / widthOfOneSecond

    if (this.timeMsElapsed >= this.timeMsOffset + ((maxHistorySeconds / 2) * 1000)) {
      this.timeMsOffset = this.timeMsElapsed

      Object
        .values(this.waves)
        .forEach((wave: Pulse) => {
          if (wave instanceof BinaryWavePulse) {
            wave.truncateSegments(screenWidth)
          }
        })
    }

    this.onTick?.(this.computeWaveGeometry(this.waves))
  }

  /**
   * Triggers all oscillation update events.
   */
  tick = (): void => {
    if (this.isPaused) return

    const now = Date.now()
    const MAX_IDLE_TIME = 10000 // 10 seconds
    let hasUpdated = false

    if (now >= this.lastUpdateTime + MAX_IDLE_TIME) {
      // if the oscilloscope has been idle for a while, then re-start it rather than try to catch up to the current time
      this.stop()
      this.start()
      return
    }

    while (now >= this.lastUpdateTime + this.refreshRate) {
      // loop through each refresh period elapsed and simulate as if they actually elapsed
      const delta = Math.min(this.refreshRate, now - this.lastUpdateTime)

      hasUpdated = true
      this.timeMsElapsed += delta
      this.lastUpdateTime += delta
      this.update()
    }

    if (hasUpdated) {
      this.broadcast()
    }

    requestAnimationFrame(this.tick)
  }

  /**
   * Updates all waves with the given tick count.
   */
  update = (): void => {
    Object
      .values(this.waves)
      .forEach((wave) => wave.update(this.timeMsElapsed))
  }

  /**
   * Adds the given wave to the oscillator instance.
   *
   * @param {Pulse} wave
   */
  add = (wave: Pulse): void => {
    if (!this.waves.hasOwnProperty(wave.id)) {
      this.waves[wave.id] = wave

      if (this.timeMsElapsed === 0) {
        this.broadcast()
      }
    }
  }

  /**
   * Removes the wave having the given id.
   *
   * @param {Pulse} wave
   */
  remove = (wave?: Pulse): void => {
    if (!wave) return

    delete this.waves[wave.id]

    if (Object.keys(this.waves).length === 0) {
      this.stop()
      this.clear()
    }

    this.broadcast()
  }
}
