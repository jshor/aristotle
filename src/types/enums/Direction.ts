enum Direction {
  Left = 0,
  Top,
  Right,
  Bottom
}

export default Direction
