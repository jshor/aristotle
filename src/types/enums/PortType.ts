enum PortType {
  Output = 0,
  Input
}

export default PortType
