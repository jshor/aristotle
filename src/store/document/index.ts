import { StoreGetters, defineStore } from 'pinia'
import { DocumentState, state } from './state'
import * as clipboardActions from './actions/clipboard'
import * as connectionActions from './actions/connection'
import * as editorActions from './actions/editor'
import * as groupingActions from './actions/grouping'
import * as itemActions from './actions/item'
import * as panningActions from './actions/panning'
import * as portActions from './actions/port'
import * as positioningActions from './actions/positioning'
import * as oscillatorActions from './actions/oscillator'
import * as rotationActions from './actions/rotation'
import * as selectionActions from './actions/selection'
import * as simulationActions from './actions/simulation'
import * as sizingActions from './actions/sizing'
import * as snappingActions from './actions/snapping'
import * as undoRedoActions from './actions/undoRedo'
import * as zIndexActions from './actions/zIndex'
import * as zoomingActions from './actions/zooming'
import * as getters from './getters'
import sortByZIndex from '@/utils/sortByZIndex'
import BaseItem from '@/types/interfaces/BaseItem'
import Item from '@/types/interfaces/Item'
import Connection from '@/types/interfaces/Connection'

export const createDocumentStore = (documentId: string) => defineStore({
  id: documentId,
  state,
  getters: {
    ...getters,

    baseItems (state) {
      return (Object
        .values(state.connections) as BaseItem[])
        .concat(Object.values(state.items) as BaseItem[])
        .sort(sortByZIndex)
    },

    hasSelection (state) {
      return state.selectedItemIds.size > 0 || state.selectedConnectionIds.size > 0
    },

    hasSelectedItems (state) {
      return state.selectedItemIds.size > 0
    },

    canGroup (state) {
      let ungroupedCount = 0
      let groupedCount = 0

      const hasUngrouped = (list: Set<string>, map: Record<string, Connection | Item>) => {
        for (const id of list) {
          map[id].groupId
            ? groupedCount++
            : ungroupedCount++

          if (ungroupedCount > 1 || (groupedCount && ungroupedCount)) {
            // at least two items are ungrouped, or at least one item is grouped and at least one item is ungrouped
            return true
          }
        }

        return false
      }

      return hasUngrouped(state.selectedItemIds, state.items) || hasUngrouped(state.selectedConnectionIds, state.connections)
    },

    canUngroup (state) {
      return state.selectedGroupIds.size > 0
    },

    canUndo (state) {
      return state.undoStack.length > 0
    },

    canRedo (state) {
      return state.redoStack.length > 0
    }
  },
  actions: {
    ...clipboardActions,
    ...connectionActions,
    ...editorActions,
    ...groupingActions,
    ...itemActions,
    ...panningActions,
    ...portActions,
    ...positioningActions,
    ...oscillatorActions,
    ...rotationActions,
    ...selectionActions,
    ...simulationActions,
    ...snappingActions,
    ...sizingActions,
    ...undoRedoActions,
    ...zIndexActions,
    ...zoomingActions
  }
})

export type DocumentStore = ReturnType<typeof createDocumentStore>

export type DocumentStoreInstance = ReturnType<DocumentStore>
