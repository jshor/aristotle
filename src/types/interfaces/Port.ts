import BinaryWavePulse from '@/store/document/oscillator/BinaryWavePulse'
import PortType from '../enums/PortType'
import Direction from '../enums/Direction'
import Point from './Point'

interface Port {
  id: string
  name: string
  elementId: string
  virtualElementId?: string
  position: Point
  type: PortType
  rotation: number
  orientation: Direction
  value: number
  hue: number
  isFreeport: boolean
  isMonitored: boolean
  connectedPortIds: string[]
  wave?: BinaryWavePulse
}

export default Port
