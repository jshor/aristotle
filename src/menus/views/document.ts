import { MenuFactory } from '@/types/interfaces/MenuFactory'
import { DocumentStore } from '@/store/document'
import { createHelpSubmenu } from '../submenus/help'
import { createEditSubmenu } from '../submenus/edit'
import { createFileSubmenu } from '../submenus/file'
import { createViewSubmenu } from '../submenus/view'
import { t } from '@/utils/i18n'

/**
 * Creates the document application menu.
 */
export const createDocumentMenu: MenuFactory = (useDocumentStore?: DocumentStore) => {
  return [
    {
      label: t('menu.file.parent'),
      submenu: createFileSubmenu(useDocumentStore)
    },
    {
      label: t('menu.edit.parent'),
      submenu: createEditSubmenu(useDocumentStore)
    },
    {
      label: t('menu.view.parent'),
      submenu: createViewSubmenu(useDocumentStore)
    },
    {
      label: t('menu.help.parent'),
      role: 'help',
      submenu: createHelpSubmenu()
    }
  ]
}
