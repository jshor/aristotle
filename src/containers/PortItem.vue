<template>
  <template v-if="connectionExperiment">
    <wire-draggable
      data-test="experiment-wire"
      :geometry="connectionExperiment.geometry"
      :style="{
        left: connectionExperiment.position.x + 'px',
        top: connectionExperiment.position.y + 'px'
      }"
      :source-value="port.value"
    />
    <port-handle
      v-if="connectionExperiment"
      data-test="experiment-port"
      :style="{
        left: connectionExperiment.targetPosition.x + 'px',
        top: connectionExperiment.targetPosition.y + 'px',
        transform: `translate(-50%, -50%)`
      }"
    />
  </template>

  <draggable
    :position="port.position"
    :style="{ zIndex }"
    @drag-start="createConnectionExperiment(port.id)"
    @drag="updateConnectionExperiment"
    @drag-end="terminateConnectionExperiment"
    @touchhold="onTouchHold"
    @dblclick="onDoubleClick"
  >
    <port-pivot>
      <port-handle
        data-test="port"
        :active="isActive"
        :hue="port.isMonitored ? port.hue : 0"
      />
    </port-pivot>
  </draggable>
</template>

<script lang="ts">
import { computed, defineComponent, PropType } from 'vue'
import PortHandle from '@/components/port/PortHandle.vue'
import PortPivot from '@/components/port/PortPivot.vue'
import WireDraggable from '@/components/basic/WireDraggable.vue'
import { DocumentStore } from '@/store/document'
import Draggable from '@/components/interactive/Draggable.vue'
import renderLayout from '@/store/document/geometry/wire'
import Port from '@/types/interfaces/Port'
import { ITEM_BASE_Z_INDEX } from '@/constants'
import Direction from '@/types/enums/Direction'

export default defineComponent({
  name: 'PortItem',
  components: {
    WireDraggable,
    PortHandle,
    PortPivot,
    Draggable
  },
  props: {
    port: {
      type: Object as PropType<Port>,
      required: true
    },

    /** Document store instance. */
    store: {
      type: Function as PropType<DocumentStore>,
      required: true
    }
  },
  setup (props, { emit }) {
    const store = props.store()
    const {
      createConnectionExperiment,
      updateConnectionExperiment,
      terminateConnectionExperiment
    } = store

    /** Whether or not the port is active (i.e., visually "large" to accept a new connection). */
    const isActive = computed(() => store.activePortId === props.port.id || store.connectablePortIds.has(props.port.id))

    /** The geometry for the connection experiment. */
    const connectionExperiment = computed(() => {
      if (store.connectionExperiment?.sourceId !== props.port.id) {
        // if this port is not part of the active connection experiment then there is no wire to render
        return null
      }

      const { targetPosition } = store.connectionExperiment
      const geometry = renderLayout(props.port, {
        position: targetPosition,
        orientation: Direction.Right,
        rotation: props.port.rotation,
        canInflect: true
      })
      const position = {
        x: Math.min(props.port.position.x, targetPosition.x) + geometry.minX,
        y: Math.min(props.port.position.y, targetPosition.y) + geometry.minY
      }

      return { geometry, position, targetPosition }
    })

    // use the same CSS z-index value for the port as its parent item
    const zIndex = computed(() => store.items[props.port.elementId]?.zIndex + ITEM_BASE_Z_INDEX)

    /**
     * Toggles oscilloscope monitoring on the port on touch-hold.
     */
    function onTouchHold ($event: TouchEvent) {
      $event.stopPropagation()
      $event.preventDefault()
      navigator.vibrate(100)

      store.togglePortMonitoring(props.port.id)
    }

    /**
     * Toggles oscilloscope monitoring on the port on double-click.
     */
    function onDoubleClick ($event: MouseEvent) {
      $event.stopPropagation()
      $event.preventDefault()

      store.togglePortMonitoring(props.port.id)
    }

    return {
      connectionExperiment,
      isActive,
      store,
      zIndex,
      onTouchHold,
      onDoubleClick,
      createConnectionExperiment,
      updateConnectionExperiment,
      terminateConnectionExperiment
    }
  }
})
</script>
