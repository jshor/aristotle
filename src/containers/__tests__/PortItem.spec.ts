import { mount, VueWrapper } from '@vue/test-utils'
import { setActivePinia, createPinia } from 'pinia'
import PortItem from '../PortItem.vue'
import { createDocumentStore, DocumentStore, DocumentStoreInstance } from '@/store/document'
import { createPort, stubAll } from '@/store/document/actions/__tests__/__helpers__'
import PortType from '@/types/enums/PortType'
import Port from '@/types/interfaces/Port'
import renderLayout from '@/store/document/geometry/wire'
import Direction from '@/types/enums/Direction'

setActivePinia(createPinia())

describe('Item container', () => {
  let wrapper: VueWrapper<typeof PortItem>
  let storeDefinition: DocumentStore
  let store: DocumentStoreInstance
  let port: Port

  beforeEach(() => {
    storeDefinition = createDocumentStore('test-id')
    store = storeDefinition()

    port = createPort('port', 'item', PortType.Input, { name: 'Test Port' })
    store.ports = { port }

    wrapper = mount(PortItem, {
      props: {
        port,
        store: storeDefinition
      }
    })
  })

  afterEach(() => {
    vi.resetAllMocks()
    store.$reset()
  })

  it('should match the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  describe('port drop activity', () => {
    it('should show the port as active when its ID is the `activePortId`', async () => {
      store.activePortId = port.id
      await wrapper.vm.$nextTick()

      expect(wrapper.find('.port-handle__display--active').exists()).toBe(true)
    })

    it('should not show the port as active when its ID differs from the `activePortId`', async () => {
      store.activePortId = 'some-other-port'
      await wrapper.vm.$nextTick()

      expect(wrapper.find('.port-handle__display--active').exists()).toBe(false)
    })

    it('should not show the port as active when there is no active port in the store', async () => {
      store.activePortId = null
      await wrapper.vm.$nextTick()

      expect(wrapper.find('.port-handle__display--active').exists()).toBe(false)
    })
  })

  describe('port monitoring', () => {
    beforeEach(() => {
      stubAll(store, ['togglePortMonitoring'])
    })

    it('should toggle port monitoring when double-clicked', async () => {
      await wrapper
        .findComponent({ name: 'Draggable' })
        .trigger('dblclick')

      expect(store.togglePortMonitoring).toHaveBeenCalledTimes(1)
      expect(store.togglePortMonitoring).toHaveBeenCalledWith(port.id)
    })

    it('should toggle port monitoring when double-clicked', async () => {
      await wrapper
        .findComponent({ name: 'Draggable' })
        .vm
        .$emit('touchhold', new TouchEvent('touchhold'))

      expect(store.togglePortMonitoring).toHaveBeenCalledTimes(1)
      expect(store.togglePortMonitoring).toHaveBeenCalledWith(port.id)
    })

    it('should show the hue of the port if it is monitored', async () => {
      store.ports[port.id].isMonitored = true
      store.ports[port.id].hue = 250

      await wrapper.vm.$nextTick()

      expect(wrapper.findComponent({ name: 'PortHandle' }).props().hue).toEqual(port.hue)
    })

    it('should set the hue of the port to 0 if it is not monitored', async () => {
      store.ports[port.id].isMonitored = false
      store.ports[port.id].hue = 250

      await wrapper.vm.$nextTick()

      expect(wrapper.findComponent({ name: 'PortHandle' }).props().hue).toEqual(0)
    })
  })

  describe('when a connection experiment is initiated', () => {
    it('should not draw a connection experiment if one is not active', async () => {
      store.connectionExperiment = null

      await wrapper.vm.$nextTick()

      expect(wrapper.find('[data-test="experiment-wire"]').exists()).toBe(false)
      expect(wrapper.find('[data-test="experiment-port"]').exists()).toBe(false)
    })

    it('should not draw a connection experiment if the current one does not belong to this port', async () => {
      const targetPosition = { x: 100, y: 200 }

      store.connectionExperiment = {
        sourceId: 'some-other-port-id',
        targetPosition
      }

      await wrapper.vm.$nextTick()

      expect(wrapper.find('[data-test="experiment-wire"]').exists()).toBe(false)
      expect(wrapper.find('[data-test="experiment-port"]').exists()).toBe(false)
    })

    it('should draw the connection experiment wire and port handle', async () => {
      const targetPosition = { x: 100, y: 200 }

      store.connectionExperiment = {
        sourceId: port.id,
        targetPosition
      }

      await wrapper.vm.$nextTick()

      const wire = wrapper.find('[data-test="experiment-wire"]')
      const handle = wrapper.find('[data-test="experiment-port"]')
      const layout = renderLayout(port, {
        position: targetPosition,
        orientation: Direction.Right,
        rotation: port.rotation,
        canInflect: true
      })

      expect(wrapper.html()).toMatchSnapshot()
      expect(wire.exists()).toBe(true)
      expect(handle.exists()).toBe(true)
      expect(wire.attributes('style')).toContain(`left: ${layout.minX}px`)
      expect(wire.attributes('style')).toContain(`top: ${layout.minY}px`)
      expect(handle.attributes('style')).toContain('left: 100px')
      expect(handle.attributes('style')).toContain('top: 200px')
    })
  })
})
