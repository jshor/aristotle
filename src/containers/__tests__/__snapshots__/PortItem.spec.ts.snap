// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item container should match the snapshot 1`] = `
"<!--v-if-->
<div class="draggable" style="left: 0px; top: 0px; z-index: NaN;" tabindex="-1">
  <!--v-if-->
  <div class="port-pivot" style="transform: rotate(-0deg);">
    <div class="port-handle" data-test="port">
      <div class="port-handle__display">
        <div class="port-handle__hue"></div>
      </div>
    </div>
  </div>
</div>"
`;

exports[`Item container when a connection experiment is initiated should draw the connection experiment wire and port handle 1`] = `
"<svg class="wire" width="180" height="230" aria-roledescription="connection" data-test="experiment-wire" style="left: -57.5px; top: -7.5px;">
  <path class="wire__outline" fill="none" transform="translate(57.5, 7.5)" d="M 0 0 C -50 0 50 200 100 200"></path>
  <path class="wire__display" fill="none" transform="translate(57.5, 7.5)" d="M 0 0 C -50 0 50 200 100 200"></path>
  <!--v-if-->
  <path class="wire__clickable" fill="none" transform="translate(57.5, 7.5)" d="M 0 0 C -50 0 50 200 100 200"></path>
</svg>
<div class="port-handle" data-test="experiment-port" style="left: 100px; top: 200px; transform: translate(-50%, -50%);">
  <div class="port-handle__display">
    <div class="port-handle__hue"></div>
  </div>
</div>
<div class="draggable" style="left: 0px; top: 0px; z-index: NaN; z-index: NaN;" tabindex="-1">
  <!--v-if-->
  <div class="port-pivot" style="transform: rotate(-0deg);">
    <div class="port-handle" data-test="port">
      <div class="port-handle__display">
        <div class="port-handle__hue"></div>
      </div>
    </div>
  </div>
</div>"
`;
